src/LineageTree/lineageTree.py

```diff
@@ -2400,6 +2403,595 @@ def get_cells_at_t_from_root(self, r: [int, list], t: int = None) -> list:
+                j -= 1
+            path.append((i, j))
+
+        # Strip infinity edges from cost_mat before returning
+        cost_mat = cost_mat[1:, 1:]
+        return (path[::-1], cost_mat)
+
+    @staticmethod
+    def __fast_dp_abs(dist_mat, w, start_d=0, back_d=0):
+        """
+        Find DTW minimum cost between two series using dynamic programming.
+        Fast algorithm using a uncentered Sakoe-Chiba band width w
+
+            Args:
+                dist_mat (matrix): distance matrix obtained by the function calculate_dtw
+                w (int): window size
+                start_d (int): start delay
+                back_d (int): end delay
+
+            Returns:
+                (tuple of tuples) Aligment path
+                (matrix) Cost matrix
+        """
+        N, M = dist_mat.shape
+
+        # Initialize the cost matrix
+        cost_mat = np.full((N + 1, M + 1), np.inf)
+
+        w_limit = max(w, abs(N - M))  # Calculate the Sakoe-Chiba band width
+
+        # Fill the cost matrix while keeping traceback information
+        traceback_mat = np.zeros((N, M))
+
+        cost_mat[: start_d + 1, 0] = 0
+        cost_mat[0, : start_d + 1] = 0
+
+        cost_mat[N - back_d :, M] = 0
+        cost_mat[N, M - back_d :] = 0
+
+        for i in range(N):
+            for j in range(M):
+                if (
+                    abs(i - j) <= w_limit
+                ):  # Only consider elements within the Sakoe-Chiba band - a[i][n-1-i]
+                    penalty = [
+                        cost_mat[i, j],  # match (0)
+                        cost_mat[i, j + 1],  # insertion (1)
+                        cost_mat[i + 1, j],
+                    ]  # deletion (2)
+                    i_penalty = np.argmin(penalty)
+                    cost_mat[i + 1, j + 1] = (
+                        dist_mat[i, j] + penalty[i_penalty]
+                    )
+                    traceback_mat[i, j] = i_penalty
+
+        min_index1 = np.argmin(cost_mat[N - back_d :, M])
+        min_index2 = np.argmin(cost_mat[N, M - back_d :])
+
+        if (
+            cost_mat[N - back_d + min_index1, M]
+            > cost_mat[N, M - back_d + min_index2]
+        ):
+            i = N - 1
+            j = M - back_d + min_index2 - 1
+        else:
+            i = N - back_d + min_index1 - 1
+            j = M - 1
+
+        path = [(i, j)]
+        while (i > start_d and j > 0) or (i > 0 and j > start_d):
+            # while i > start_d or j > 0:
+            tb_type = traceback_mat[i, j]
+            if tb_type == 0:
+                # Match
+                i -= 1
+                j -= 1
+            elif tb_type == 1:
+                # Insertion
+                i -= 1
+            elif tb_type == 2:
+                # Deletion
+                j -= 1
+            path.append((i, j))
+
+        # Strip infinity edges from cost_mat before returning
+        cost_mat = cost_mat[1:, 1:]
+        return (path[::-1], cost_mat)
```

Comment:
There is a lot of repeated code between the three version of `dp`, maybe you could refactor?